utils/pubsub.go

```diff
@@ -14,56 +14,109 @@ type DataEvent struct {
 type DataChannel chan DataEvent
 
-// DataChannelSlice is a slice of DataChannels
-type DataChannelSlice []DataChannel
 type PublishSubscriber interface {
 	Publish(topic string, data interface{})
-	PublishToChannel(channel DataChannel, topic string, data interface{})
 	Subscribe(topic string, ch DataChannel)
 }
 
 // EventBus stores the information about subscribers interested for a particular topic
 type EventBus struct {
-	subscribers map[string]DataChannelSlice
-	rm          sync.RWMutex
+	lastEventMutex sync.RWMutex
+	// lastEvent holds the last event for each topic so that we can send it to new subscribers
+	lastEvent map[string]*DataEvent
+
+	subscribersMutex sync.RWMutex
+	// subscribers keep the list of subscription publishers per topic
+	subscribers map[string]publisherSlice
 }
 
 func (eb *EventBus) Publish(topic string, data interface{}) {
-	eb.rm.RLock()
-	if chans, found := eb.subscribers[topic]; found {
-		// this is done because the slices refer to same array even though they are passed by value
-		// thus we are creating a new slice with our elements thus preserve locking correctly.
-		// special thanks for /u/freesid who pointed it out
-		channels := append(DataChannelSlice{}, chans...)
-		go func(data DataEvent, dataChannelSlices DataChannelSlice) {
-			for _, ch := range dataChannelSlices {
-				//Publishing to channels in separate goroutines
-				go func(ch DataChannel, data DataEvent) {
-					ch <- data
-				}(ch, data)
-			}
-		}(DataEvent{Data: data, Topic: topic}, channels)
+	eb.subscribersMutex.RLock()
+	defer eb.subscribersMutex.RUnlock()
+	eb.lastEventMutex.Lock()
+	defer eb.lastEventMutex.Unlock()
+
+	evt := &DataEvent{Data: data, Topic: topic}
+	if eb.lastEvent == nil {
+		eb.lastEvent = map[string]*DataEvent{}
+	}
+	eb.lastEvent[topic] = evt
+
+	if publishers, found := eb.subscribers[topic]; found {
+		for _, publisher := range publishers {
+			publisher.publish(evt)
+		}
 	}
-	eb.rm.RUnlock()
-}
 
-func (eb *EventBus) PublishToChannel(channel DataChannel, topic string, data interface{}) {
-	eb.rm.RLock()
-	go func() {
-		channel <- DataEvent{Data: data, Topic: topic}
-	}()
-	eb.rm.RUnlock()
 }
 
 func (eb *EventBus) Subscribe(topic string, ch DataChannel) {
-	eb.rm.Lock()
+	eb.subscribersMutex.Lock()
+	defer eb.subscribersMutex.Unlock()
+	eb.lastEventMutex.RLock()
+	defer eb.lastEventMutex.RUnlock()
+
+	p := &publisher{channel: ch}
 	if prev, found := eb.subscribers[topic]; found {
-		eb.subscribers[topic] = append(prev, ch)
+		eb.subscribers[topic] = append(prev, p)
 	} else {
 		if eb.subscribers == nil {
-			eb.subscribers = map[string]DataChannelSlice{}
+			eb.subscribers = map[string]publisherSlice{}
 		}
-		eb.subscribers[topic] = append([]DataChannel{}, ch)
+		eb.subscribers[topic] = publisherSlice{p}
+	}
+	if eb.lastEvent[topic] != nil {
+		p.publish(eb.lastEvent[topic])
+	}
+}
+
+// publisherSlice is a slice of publisher pointers
+type publisherSlice []*publisher
```

Comment:
[Minor]

Is this publisherSlice or subscribersSlice?

Reply:
You are right that the terms used here are a bit confusing, assigning a `publisherSlice` to a `subscribers` variable doesn't seem to be logical! I will revisit them!